P1684-mdarray/P1684.md

```diff
@@ -644,6 +651,504 @@ The larger or equal is intentional, because the mapping may already be not exhau
+```c++
+std::vector<float>&& f0(std::vector<float>&& storage)
+{
+  mdspan m0{storage.data(), 2, 3, 4};
+  // ... compute with m0 ...
+  return std::move(storage);
+}
+
+std::vector<float>&& f1(std::vector<float>&& storage)
+{
+  mdspan m1{storage.data(), 6, 4};
+  // ... compute with m1 ...
+  return std::move(storage);
+}
+
+int main()
+{
+  // Only one vector allocation, created as a temporary,
+  // and moved into used_storage.
+  vector<float> used_storage(f1(f0(vector<float>(24, 0.0f))));
+  // ... use used_storage for other things ...
+  return 0;
+}
+```
+
+Here are some important points in the above example.
+
+1. `mdspan` views the elements in `storage`.
+    It does *not* take ownership of or even store the container.
+
+2. The above example only passes the `vector` along
+    in order to look as similar as possible
+    to the first `f0`, `f1` example.
+    The functions here do not *need* to pass the `vector` along;
+    they just need to make sure that the `vector`'s lifetime
+    contains the lifetimes of all the `mdspan` that view its elements.
+
+#### "Reshape" reuses storage and elements
+
+Reusing storage and elements
+is equivalent to the "reshape" facility
+found in other programming languages and libraries.
+This facility reinterprets an existing array
+as an array with a different (multidimensional) "shape."
+Examples include Fortran's `RESHAPE` intrinsic function.
+The `mdarray` class cannot be "reshaped in place,"
+because the extents and layout are part of the type.
+Even if all the extents are dynamic, the *rank* is static.
+
+If `mdarray` had a facility for extracting the container,
+then one way to implement an "out-of-place reshape"
+would be to extract the container from the source `mdarray`
+and move it into a new destination `mdarray`.
+However, depending on the container type,
+this may copy the elements instead of viewing the original elements.
+An example of such a container type is `std::array`.
+
+Thus, the only way to implement a reshape operation
+that *always* views the original elements
+would be to use `mdspan`.  That is, the reshape operation
+could view an existing `mdarray` as an `mdspan`
+with different extents and/or layout.
+
+### Effects of extracting the container from an mdarray
+
+The discussion above explains why we do not think
+`mdarray` needs an operation for extracting its container.
+If we were nevertheless to add such an operation,
+this section explains its postconditions:
+that is, the state of the "extracted-from" `mdarray`.
+
+#### An "extracted-from" mdarray is like a moved-from object
+
+The hypothetical `extract_container()` member function
+would have the effect of returning `move(ctr_)`.
+If the caller uses the result in a `container_type`
+move constructor or move assignment operator,
+then this would leave the `mdarray`'s container in a moved-from state.
+Given that `mdarray` can accept any contiguous container
+and imposes no requirements on the container's moved-from state,
```

Comment:
@nliber This is a good point.  I really wish C++ had a "sink state" that the compiler could enforce, instead of whatever this moved-from state is supposed to be.  Here's a summary of my thoughts.

1. P1684 doesn't currently impose any requirements on the container's moved-from state, neither explicitly nor implicitly.  The move constructor and move assignment operator are both `= default`.  Thus, I have to conclude that a moved-from mdarray has a "valid but unspecified" container, a thing completely unusable for anything other than destruction.

2. Extraction is worse because it's not syntactically marked.  It's just some member function.  On the other hand, we could syntactically mark it by making it rvalue-qualified (`&&`).

3. If the container is `vector` or `vector`-like, it would be nice if returning a (named) mdarray from a function didn't double-allocate and copy.